examples/LoRaWAN/LoRaWAN_End_Device_Reference/LoRaWAN_End_Device_Reference.ino

```diff
@@ -173,44 +138,45 @@ void loop() {
   // retrieve the last uplink frame counter
   uint32_t fcntUp = node.getFcntUp();
 
-  Serial.print(F("[LoRaWAN] Sending uplink packet ... "));
-  String strUp = "Hello!" + String(fcntUp);
+  Serial.print(F("[LoRaWAN] Sending uplink packet #"));
+  Serial.println(fcntUp);
+  String strUp = "Hello! " + String(fcntUp);
   
   // send a confirmed uplink to port 10 every 64th frame
   // and also request the LinkCheck and DeviceTime MAC commands
   if(fcntUp % 64 == 0) {
+    Serial.print(F("[LoRaWAN] Requesting LinkCheck and DeviceTime"));
     node.sendMacCommandReq(RADIOLIB_LORAWAN_MAC_LINK_CHECK);
     node.sendMacCommandReq(RADIOLIB_LORAWAN_MAC_DEVICE_TIME);
     state = node.uplink(strUp, 10, true);
   } else {
     state = node.uplink(strUp, 10);
   }
-  if(state == RADIOLIB_ERR_NONE) {
-    Serial.println(F("success!"));
-  } else {
+  if(state != RADIOLIB_ERR_NONE) {
     Serial.print(F("failed, code "));
     Serial.println(state);
   }
 
-  // after uplink, you can call downlink(),
-  // to receive any possible reply from the server
-  // this function must be called within a few seconds
-  // after uplink to receive the downlink!
-  Serial.print(F("[LoRaWAN] Waiting for downlink ... "));
+  // after uplink, you must call downlink() to receive any possible reply
+  // from the server. This function must be called before the Rx1 delay
+  // for the network. Typically this is 5s after end of uplink.
+  Serial.println(F("[LoRaWAN] Waiting for downlink ... "));
   String strDown;
 
-  // you can also retrieve additional information about 
-  // uplink or downlink by passing a reference to
-  // LoRaWANEvent_t structure
-  LoRaWANEvent_t event;
-  state = node.downlink(strDown, &event);
+  // you can also retrieve additional information about an uplink or 
+  // downlink by passing a reference to LoRaWANEvent_t structure
+  LoRaWANEvent_t downlinkDetails;
+  state = node.downlink(strDown, &downlinkDetails);
   if(state == RADIOLIB_ERR_NONE) {
-    Serial.println(F("success!"));
-
-    // print data of the packet (if there are any)
+    // print data of the packet
     Serial.print(F("[LoRaWAN] Data:\t\t"));
     if(strDown.length() > 0) {
-      Serial.println(strDown);
+      for (uint8_t c = 0; c < strDown.length(); c++) {
+        uint8_t value = strDown[c];
+        if (value < 10) Serial.print(F("0"));
+        Serial.print(value, HEX);
+      }
+      Serial.println();
```

Comment:
Bit conflicted over this - on TTN sending text is not just frowned on but actually nuked from high orbit.

The problem here is that, as we've recently seen, people don't type the hex for "Hello" in to the downlink messages on the console - so they don't understand what they are getting when it arrives.

The noobs version is all byte array both on send & receive. I'll change these two examples to using all byte array for comment.

Reply:
> The problem here is that, as we've recently seen, people don't type the hex for "Hello" in to the downlink messages on the console - so they don't understand what they are getting when it arrives.

In that case, it would make more sense to remove the `String` from here completely and just use byte arrays - maybe with a comment that `String` is also available (but not preferred). I'm definitely no fan of Arduino's `String` class.

Reply:
I don't mean to interject out of the blue in this code review. But my 2ct is that people have a tendency to send String messages over LoRa..not sure really why that is. I am suspecting it's partly ignorance of what a low bandwidth network protocol is.
I just went through the same exercise in the opensource project I help maintain (ClusterDuck Protocol). Literally banning the use of `String` as message payload and requiring now that all messages be provided as a byte array.

We then provide an example of how you can convert back into a string for (better) readability..or if you are too worried you can't read hex values fluently.

Anyway..I love reading these code reviews on RadioLib..love the project, so keep it up!

Reply:
Hi @amirna2, the one that puzzles me is when someone sends text for a float and then has to figure out how to parse it out at the other end! The worst I've seen is a honking great JSON string that got too large even with single character field names!

Reply:
Yeah. I didn't want to mention the JSON string..but I have seen that as well. The reason to stick to binary is also to comply with regulations and on-air time requirement that exist on public network.